src/System.Management.Automation/engine/ParameterBinderBase.cs

```diff
@@ -1258,7 +1258,35 @@ internal InternalCommand Command
                         bindingTracer.WriteLine(
                             "CONVERT arg type to param type using LanguagePrimitives.ConvertTo");
 
-                        result = LanguagePrimitives.ConvertTo(currentValue, toType, CultureInfo.CurrentCulture);
+                        // If we are in constrained language mode and the target command is trusted, which is often
+                        // the case for C# cmdlets, then we allow type conversion to the target parameter type.
+                        //
+                        // However, we don't allow Hashtable-to-Object conversion (PSObject and IDictionary) because
+                        // those can lead to property setters that probably aren't expected. This is enforced by
+                        // setting 'Context.LanguageModeTransitionInParameterBinding' to true before the conversion.
+                        bool changeLanguageModeForTrustedCommand = 
+                            Context.LanguageMode == PSLanguageMode.ConstrainedLanguage &&
+                            this.Command.CommandInfo.DefiningLanguageMode == PSLanguageMode.FullLanguage;
+                        bool oldLangModeTransitionStatus = Context.LanguageModeTransitionInParameterBinding;
```

Comment:
This diff doesn't look right.  There is a new ExecutionContext.LanguageModeTransitionInParameterBinding property but this doesn't show where it is defined or used.

Reply:
Please disregard as I was comparing to PowerShell Core6 code.  I now see this defined in WindowsPowerShell